src/aws_utils/aio_npy_storage.py

```diff
@@ -0,0 +1,224 @@
+import functools
+import logging
+import aioboto3
+import aiofiles
+import json
+import numpy as np
+from pathlib import Path
+from typing import Any, AsyncGenerator
+
+from .npy_storage import METADATA_KEY
+
+
+logger = logging.getLogger(__name__)
+
+
+@functools.lru_cache
+async def _get_async_client():
+    async with aioboto3.Session().client("s3") as client:
+        return client
+
+
+class AIONpyReader:
+    """
+    For asyncio-based reading of numpy arrays from S3 with optional local caching.
+    Uses memory efficient chunk-based reading using HTTP range requests.
+    """
+
+    def __init__(
+        self, bucket: str, key_prefix: str, cache_dir: str | None = None, client=None
+    ):
+        self.bucket = bucket
+        self.key_prefix = key_prefix.rstrip("/")
+        self._metadata_cache: dict[str, Any] | None = None
+        self._shard_metadata: list[dict[str, Any]] = []
+        self._cache_locally = cache_dir is not None
+        self._local_cache_dir = Path(cache_dir) if cache_dir else None
+        self._metadata_key = f"{self.key_prefix}/{METADATA_KEY}"
+        self._client = client or _get_async_client()
+
+    def _get_shard_key(self, shard_filename: str) -> str:
+        return f"{self.key_prefix}/{shard_filename}"
+
+    async def prewarm_cache(self) -> None:
+        """
+        Preload the local cache with the metadata and shards.
+        """
+        if not self._cache_locally:
```

Comment:
Could consider raising an exception here, since calling code shouldn't be using this method if cache is disabled.